### ISSUE TITLE
one of the variables needed for gradient computation has been modified by an inplace operation

### ISSUE THREAD

**Issue by woaipichuli** [pytorch/pytorch#36470]
I found pytorch returns error report when running codes below. I attempted to realize calculation which can be formulated as f(n) = Sigmoid(f(n-1)*x).

```
x = torch.ones(1)
x.requires_grad_()
a = 1
f = torch.zeros(10)
f[0] = a
for i in range(1,10):
    f[i] = f[i-1]*x
    f[i] = 1/(1.0+torch.exp(-f[i]))
loss = torch.sum(f)
grad = torch.autograd.grad(loss,[x])[0]
```

    As can be seen, results of f(n) will not be modified after the corresponding calcualtion. I considered we may not reuse the tensor for  intermediate results. Thus, I rewrite the code and it work. I wonder whether it is necessary as this is obviously not efficient.

```
x= torch.ones(1)
x.requires_grad_()
a = 1
f = torch.zeros(10)
f[0] = a
for i in range(1,10):
    tmp = a
    for j in range(i):
        tmp = 1/(1.0+torch.exp(-(a*x)))
    f[i] = tmp
loss = torch.sum(f)
grad = torch.autograd.grad(loss,[x])[0]
```

**Comment by izdeby** [comment]
Please, use our [forum](https://discuss.pytorch.org/) for questions. Thank you
